Outline potentially undefined globals during lowering

Keno opened the pull request:
Currently [1] it is illegal [2] in IRCode to have a GlobalRef in value position that could potentially throw. This is because in IRCode, we want to assign flags to every statement and if there are multiple things with effects in a statement, we lose precision in tracking which they apply to. However, we currently do allow this in `CodeInfo`. Now that we're starting to make more use of flags in inference also, this is becoming annoying (as it did for IRCode), so I would like to do this transformation earlier. This is an attempt to do this during lowering. It is not entirely clear that this is precisely the correct place for it. We could alternatively consider doing it during the global resolve pass in method.c, but that currently does not renumber SSAValues, so doing it during the renumbering inside lowering may be easier.

N.B.: This is against #51853, because this needs some of the inference precision improvements in that PR to avoid regressing the try/catch elision tests (which before that PR, we were incorrectly computing effects for statement-position GlobalRefs).

[1] https://github.com/JuliaLang/julia/commit/39c278b728deb04c3a32d70e3e35dcef7822c0c0
[2] https://github.com/JuliaLang/julia/blob/2f63cc99fb134fb4adb7f11ba86a4e2ab5adcd48/base/compiler/ssair/verify.jl#L54-L58

Keno committed:
Properly guard UpsilonNode unboxed store

In #51852, we are coercing a boxed `Union{@NamedTuple{progress::String}, @NamedTuple{progress::Float64}}`
to `@NamedTuple{progress::String}` via convert_julia_type. This results in a jl_cgval_t that has
a Vboxed that points to a boxed `@NamedTuple{progress::Float64}` but with a `@NamedTuple{progress::String}`
type tag that the up upsilonnode code then tries to unbox into the typed PhiC slot. This ends up treating
the Float64 as a pointer and crashing in GC. Avoid this by adding a runtime check that the converted value
is actually compatible (we already had this kind of check for the non-boxed cases) and then making the
unboxing runtime-conditional on the type of the union matching the type of the phic.

Fixes #51852

Keno committed:
WIP: Outline potentially undefined globals during lowering

Currently [1] it is illegal [2] in IRCode to have a GlobalRef in value
position that could potentially throw. This is because in IRCode,
we want to assign flags to every statement and if there are multiple
things with effects in a statement, we lose precision in tracking
which they apply to. However, we currently do allow this in `CodeInfo`.
Now that we're starting to make more use of flags in inference also,
this is becoming annoying (as it did for IRCode), so I would like
to do this transformation earlier. This is an attempt to do this
during lowering. It is not entirely clear that this is precisely
the correct place for it. We could alternatively consider doing
it during the global resolve pass in method.c, but that currently
does not renumber SSAValues, so doing it during the renumbering
inside lowering may be easier.

[1] https://github.com/JuliaLang/julia/commit/39c278b728deb04c3a32d70e3e35dcef7822c0c0
[2] https://github.com/JuliaLang/julia/blob/2f63cc99fb134fb4adb7f11ba86a4e2ab5adcd48/base/compiler/ssair/verify.jl#L54-L58

JeffBezanson commented on src/ast.c:
```diff
@@ -161,6 +161,17 @@ static value_t fl_defined_julia_global(fl_context_t *fl_ctx, value_t *args, uint
+    jl_ast_context_t *ctx = jl_ast_ctx(fl_ctx);
+    jl_sym_t *var = jl_symbol(symbol_name(fl_ctx, args[0]));
+    jl_binding_t *b = jl_get_module_binding(ctx->module, var, 0);
+    return b != NULL ? fl_ctx->T : fl_ctx->F;
```
```suggestion
    return b != NULL && b->value != NULL ? fl_ctx->T : fl_ctx->F;
```

JeffBezanson commented on base/reflection.jl:
```diff
@@ -2081,6 +2081,8 @@ function bodyfunction(basemethod::Method)
                         return nothing
                     end
+                elseif isa(fsym, Core.SSAValue)
+                    fsym = ast.code[fsym.id]
```
We might want to add the ability to annotate uses as nothrow, for auto-generated globals we know at lowering time will be defined before they're used.

Keno commented on src/ast.c:
```diff
@@ -161,6 +161,17 @@ static value_t fl_defined_julia_global(fl_context_t *fl_ctx, value_t *args, uint
+    (void)tosymbol(fl_ctx, args[0], "nothrow-julia-global");
+    jl_ast_context_t *ctx = jl_ast_ctx(fl_ctx);
+    jl_sym_t *var = jl_symbol(symbol_name(fl_ctx, args[0]));
+    jl_binding_t *b = jl_get_module_binding(ctx->module, var, 0);
```
Doesn't that cause the binding to resolve? Currently merely lowering a function that mentions a symbol does not resolve that binding in case you want to define it to something else later.

vtjnash replied:
alright, `jl_get_binding_if_bound` then

vtjnash commented on test/compiler/inference.jl:
```diff
@@ -2998,6 +2998,8 @@ end
 
 # issue #28279
+# ensure that lowering doesn't move these into statement position, which would require renumbering
+using Base: +, -
```
This state dependence, where we give very different lowering depending on whether something has been used already in the module, seems very ripe for causing future heisenbugs

Keno replied:
I agree, but on the other hand, this is supposed to be a storage optimization without semantic effect (though of course there can be bugs).

vtjnash replied:
True. I guess in the future we may want to optimize cases like this more aggressively pre-inference too. I just worry about cases where people may pre-lower the code, then ship it off to another process to execute which as slightly different globals at the time. I think Jeff had some code where he was planning to do that and Distributed.jl does that for remotecall.

Keno committed:
interpreter: Fix stacktrace for global incorrectly assumed to be in phi block

We don't fully set up the interpreter state when evaluating the phi block,
because we expect all statements to be valid in value-position (using
the IR definition of value position). However, the interpreter was
overapproximating the size of the phi-block, leading to the possibility
that the evaluation may throw. Correct that by truncating the phi
block to the actual last phi. Will be tested in existing tests after
#51970.

Keno committed:
WIP

Keno commented:
Looks like the sysimage size increase is about 1%. We could claw some of that back with more precise never-undef analysis, but I'm not sure it's worth it at this stage. If lowering will get rewritten in julia eventually anyway, we might as well do it then, since we have a much better algorithm for this in julia already.

Keno committed:
WIP

Keno committed:
WIP

Keno commented:
I think this is largely complete. However, we should merge #51990 and #51991 first and then rebase this.

vtjnash commented:
If you have a chance, how many megabytes larger is that, just for reference? It seems okay, and especially so that it is worthwhile

Keno commented:
1.5MB

Keno committed:
Fix various misc checks

oscardssmith commented:
Is this ready to merge?

JeffBezanson committed:
fix some indenting

JeffBezanson commented on src/julia-syntax.scm:
```diff
@@ -4252,10 +4252,13 @@ f(x) = yt(x)
       (and (pair? e)
-           (memq (car e) '(quote inert top core globalref outerref
+           (memq (car e) '(quote inert top core
                                  slot static_parameter)))))
```
Should we remove `static_parameter` from here since it can also throw?

Keno replied:
Hmm, I suppose, but for some reason I always see static parameters getting moved out anyway.

JeffBezanson commented on src/julia-syntax.scm:
```diff
@@ -4252,10 +4252,13 @@ f(x) = yt(x)
 (define (valid-ir-argument? e)
-  (or (simple-atom? e) (symbol? e)
+  (or (simple-atom? e)
+      (and (outerref? e) (nothrow-julia-global (cadr e)))
```
`nothrow-julia-global` should be able to accept a globalref as well.

JeffBezanson commented on src/julia-syntax.scm:
```diff
@@ -4252,10 +4252,13 @@ f(x) = yt(x)
+  (or (simple-atom? e)
+      (and (outerref? e) (nothrow-julia-global (cadr e)))
       (and (pair? e)
-           (memq (car e) '(quote inert top core globalref outerref
```
globalref and outerref should be added to the allow-list in `valid-ir-rvalue?`

JeffBezanson replied:
And maybe `valid-ir-return?` ? Not sure if that is ok though.

JeffBezanson commented on src/julia-syntax.scm:
```diff
@@ -4408,48 +4411,54 @@ f(x) = yt(x)
+                        (cons (if (and
+                                   ;; Even if we have side effects, we know that singly-assigned
+                                   ;; locals cannot be affected them, so we can inline them anyway.
+                                   (or simple? (single-assign-var? aval))
```
I think this `or` should also accept simple-atom, quote, inert, top and core? (Everything in valid-ir-argument except globals, slots, and static_parameter.)

JeffBezanson replied:
Tried it and it shaves ~450k off the sysimage.

Keno replied:
Yes, that seems reasonable

Keno committed:
Fix rebase

Keno committed:
Address review comments

Keno committed:
Add GC root annotation

Keno committed:
Update doctest

oscardssmith committed:
Merge branch 'master' into kf/outlineglobals
